File: exercise/leetcode/0146/0146.py

class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prv = None
        self.nxt = None


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = Node(0, 0)
        self.end = Node(0, 0)
        self.head.nxt = self.end
        self.end.prv = self.head
        self.table = {}

    def get(self, key: int) -> int:
        if key in self.table:
            value = self.table[key].val
            self.move_to_head(self.table[key])
            return value
        else:
            return -1
            
    def put(self, key: int, value: int) -> None:
        if key not in self.table:
            if len(self.table) >= self.capacity:
                tg_node = self.end.prv
                del self.table[tg_node.key]
                tg_node.key = key
                tg_node.val = value
                self.table[key] = tg_node
                self.move_to_head(tg_node)
            else:
                tg_node = Node(key, value)
                self.insert_to_head(tg_node)
                self.table[key] = tg_node
        else:
            tg_node = self.table[key]
            tg_node.val = value
            self.move_to_head(tg_node)

    def insert_to_head(self, node):
        t = self.head.nxt
        self.head.nxt = node
        node.prv = self.head
        node.nxt = t
        t.prv = node    
        
    def move_to_head(self, node):
        pre_node = node.prv
        nxt_node = node.nxt
        
        # remove node in linklist
        pre_node.nxt = nxt_node
        nxt_node.prv = pre_node
        
        # insert to head
        self.insert_to_head(node)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: exercise/leetcode/0146/0146.cc
#include <list>
#include <iostream>
#include <unordered_map>
#include <utility>
using namespace std;

struct Node {
    int key, val;
    Node* next = nullptr;
    Node* prev = nullptr;
    Node(int key, int val): key(key), val(val) {}
};

class LRUCache {
    Node* head_ = nullptr;
    Node* end_ = nullptr;
    unordered_map<int, Node*> table_;
    int capacity_;
public:
    LRUCache(int capacity) {
        head_ = new Node(0, 0);
        end_ = new Node(0, 0);
        head_->next = end_;
        end_->prev = head_;
        capacity_ = capacity;
    }
    
    int get(int key) {
        if (table_.find(key) == table_.end()) return -1;
        Node* tg_node = table_[key];
        remove_node_from_linklist(tg_node);
        insert_head(tg_node);
        return tg_node->val;
    }
    
    void put(int key, int value) {
        Node* tg_node = nullptr;
        if (table_.find(key) == table_.end()) {
            if (table_.size() >= capacity_) {
                tg_node = end_->prev;
                remove_node_from_linklist(tg_node);
                table_.erase(tg_node->key);
                delete tg_node;
                tg_node = nullptr;
            }
            tg_node = new Node(key, value);
            insert_head(tg_node);
            table_[key] = tg_node;
        } else {
            tg_node = table_[key];
            tg_node->val = value;
            move_to_head(tg_node);
        }
    }

    void remove_node_from_linklist(Node* node) {
        Node* pre_node = node->prev;
        Node* nxt_node = node->next;
        pre_node->next = nxt_node;
        nxt_node->prev = pre_node;
    }

    void insert_head(Node* node) {
        Node* t = head_->next;
        head_->next = node;
        node->prev = head_;
        node->next = t;
        t->prev = node;
    }

    void move_to_head(Node* node) {
        remove_node_from_linklist(node);
        insert_head(node);
    }

};
